history-network.md

```diff
@@ -147,14 +147,38 @@ each receipt/transaction and re-rlp-encode it, but only if it is a legacy transa
 
 #### Block Header
 
+
+```
+# Content types
+
+AccumulatorProof = Vector[Bytes32, 15]
+
+BlockHeaderProof = Union[None, AccumulatorProof]
```

Comment:
This is going to cause some forward compatibility issues.  Once we add `PostMergeAccumulatorProof` we will want to get rid of the `None` to make sure that proofs are required, but that will change the overall serialization scheme and effectively invalidate all of the previous "stuff"....

I think that we can just take out the Union, and send empty lists for post-merge headers *or* simply bite the bullet and work to figure out producing these proofs as well.  They are both going to be SSZ proofs... so they in theory are going to both be lists of 32 byte values.  I think the only difference is going to be in verifying the proofs...

Reply:
Yes, this is what I meant with the issue in https://github.com/ethereum/portal-network-specs/issues/175#issuecomment-1342827999

I mentioned there that we could leave the `None` in there and just not support it any more. But I agree that this is quite ugly (especially for a new specification).

Your suggestion sounds better. We can indeed define a List that we keep empty and ignore or a vector with 1 or so. Later content would have to be purged/reloaded for those headers though.

I have been playing around with those proofs here by the way: https://github.com/status-im/nimbus-eth1/pull/1376/files

Reply:
If later content will need to be purged anyway, I'm not sure I see any major difference. Maybe we could change this type to `Union[None, AccumulatorProof, PostMergeAccumulatorProof]`, allowing to experiment in the meanwhile with new proof formats, but still supply the network with proof-less post-merge headers (`None`).

Re. purging content, it is on my todo list to write a trin feature that will iterate the db and purge all invalid content, so purging data that's no longer validly encoded shouldn't be a huge blocker (after we get the feature written of course).

Reply:
I don't have a strong opinion so feel free to move forward with whatever solution you guys want.

Reply:
To summarize, the options I see:

Option I (current):
```Python
AccumulatorProof = Vector[Bytes32, 15]

# Union can be adjusted later on by adding `HistoricalRootsProof` or whatever we end up using
# None might remain viable, and if it doesn't, keep it in Union but ignore it.
# Post merge headers that are currently stored with a None, need to be purged and replaced
# in the future
# In the case that we really want get rid of the None in the content key type, all headers
# need to purged & replaced
BlockHeaderProof = Union[None, AccumulatorProof]

BlockHeaderWithProof = Container[
  header: ByteList, # RLP encoded header in SSZ ByteList
  proof: BlockHeaderProof
]
```

Option II:
```Python
AccumulatorProof = Vector[Bytes32, 15]

# Temporary list that would be empty right now. To be changed to the proof type
# we end up using. Post merge headers that are currently stored with this need to
# be purged and replace in the future
PlaceHolderProof = List[Bytes32, max_length=1] 

BlockHeaderProof = Union[AccumulatorProof, PlaceHolderProof]

BlockHeaderWithProof = Container[
  header: ByteList, # RLP encoded header in SSZ ByteList
  proof: BlockHeaderProof
]
```

Option III:

```Python
AccumulatorProof = Vector[Bytes32, 15]

# Yet another Union, which can then become Union[None, HistoricalRootsProof] or just
# `HistoricalRootsProof` if `None` isn't needed.
PlaceHolderProof = Union[None]

BlockHeaderProof = Union[AccumulatorProof, PlaceHolderProof]

BlockHeaderWithProof = Container[
  header: ByteList, # RLP encoded header in SSZ ByteList
  proof: BlockHeaderProof
]
```


Option II and Option III are very similar in the sence that `PlaceHolderProof` can be replaced to whatever without touching pre-merge headers. And post merge headers would need to be altered anyhow as proofs eventually need to be added.

I have a slight preference over option I as it is simple and represents best the current state. And in my experience, whenever trying to do something "future proof", you typically end up selecting the wrong thing and still have to make adjustments.
Option III looks best combination of current state and most future flexible but the Union in Union (can be adjusted later) is a bit bleh, so I think I rather go for Option II then if we don't want the initial `None`.
I do think that the header without a proof (`None`) is likely to be a case (as explained in previous comment). But in that scenario, we might have to do more thorough changes as data would need to get altered from without proof -> with proof. So perhaps some database resets (due to content keys?) will be required anyhow. (Also, the purging/reset of data can be a good test for the network).

Reply:
I'm also leaning towards option I for simplicity. Imo, the expectation that we will never want/have to do a network/wide database purge is unrealistic, given that the specs aren't finalized. In which case it seems ok to allow for that possibility in designing these types.

Reply:
:+1:  for option 1)